File: warrior/src/main/java/com/dumbdogdiner/warrior/gui/ArenaGUI.java
package com.dumbdogdiner.warrior.gui;

import com.dumbdogdiner.stickyapi.bukkit.gui.ClickableSlot;
import com.dumbdogdiner.stickyapi.bukkit.gui.GUI;
import com.dumbdogdiner.warrior.Warrior;
import com.dumbdogdiner.warrior.api.translation.Symbols;
import com.dumbdogdiner.warrior.api.user.WarriorUser;
import com.dumbdogdiner.warrior.api.arena.Arena;
import com.dumbdogdiner.warrior.api.builders.ItemBuilder;
import com.dumbdogdiner.warrior.api.sessions.ArenaSession;
import com.dumbdogdiner.warrior.api.translation.Constants;
import com.dumbdogdiner.warrior.managers.ArenaManager;
import com.dumbdogdiner.warrior.managers.PlayerManager;
import com.dumbdogdiner.warrior.api.util.TranslationUtil;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public class ArenaGUI extends GUI {

    /**
     * Create a GUI with the given number of rows.
     *
     * @param rows   The number of rows in the inventory, must be between 1 and 6
     * @param name   The name of the GUI, displayed as the inventory's name.
     * @param plugin The plugin responsible for this GUI
     */
    public ArenaGUI(int rows, @NotNull String name, @NotNull Plugin plugin) {
        super(rows, name, plugin);
    }

    @Override
    protected void onInventoryOpen(@NotNull InventoryOpenEvent event) {
        ItemStack placeholder = new ItemBuilder(Material.BLACK_STAINED_GLASS_PANE)
                                    .setName(" ").build();

        ItemStack blank = new ItemBuilder(Material.GRAY_DYE)
                .setName("§7Arena §3???").build();

        for(int i = 0; i < 9; i++) {
            addSlot(i, 0, placeholder);
            addSlot(i, 4, placeholder);
        }

        ClickableSlot exit = new ClickableSlot(Material.BARRIER, 1, "§c§lClose Menu", 4, 4);
        addSlot(exit);

        int arenaCount = ArenaManager.getArenas().size();
        int i = 0;

        for (int y = 1; y < 4; y++) {
            for (int x = 1; x < 8; x++) {
                if (i >= arenaCount) {
                    addSlot(x, y, blank);
                    continue;
                }
                Arena a = ArenaManager.getArenas().get(i);
                int avgRating = (int) a.getMetadata().averageRating();
                List<String> colorDesc = Arrays.stream(a.getMetadata().getDesc())
                        .map(str -> "§7§o" + str)
                        .collect(Collectors.toList());

                ItemStack arenaItem = new ItemBuilder(Material.FILLED_MAP)
                                        .setName("§7Arena §3" + a.getName())
                                        .setLore(" ", "&7Rating &8» &b" + calcRating((int) a.getMetadata().averageRating()), " ")
                                        .appendLore(colorDesc.toArray(String[]::new))
                                        .build();
                ClickableSlot cs = new ClickableSlot(arenaItem, x, y);
                addSlot(cs);
                i++;
            }
        }
    }

    private String calcRating(int rating) {
        String stars = String.valueOf(Symbols.BLACK_STAR).repeat(rating);

        return stars + String.valueOf(Symbols.WHITE_STAR).repeat(5 - rating);
    }

    @Override
    protected void onInventoryClick(@NotNull InventoryClickEvent event, @Nullable String tag) {
        ItemStack item = event.getCurrentItem();
        if(item == null) return;
        String name = item.getItemMeta().getDisplayName();

        if(name.contains("Close Menu")) {
            event.getWhoClicked().closeInventory();
        } else if(name.contains("Arena")) {
            String arenaName = name.replace("§7Arena §3", "");
            WarriorUser user = PlayerManager.get(event.getWhoClicked().getUniqueId());

            if(user == null) return;
            Arena a = ArenaManager.get(arenaName);
            if(a == null) return;

            user.setSession(new ArenaSession(user.getUserId(), a));
            String msg = Warrior.getTranslator().translate(Constants.Lang.ARENA_TELEPORT, new HashMap<>() {
                {
                    put("ARENA", a.getName());
                }
            }, user);

            user.sendMessage(TranslationUtil.getPrefix() + msg);
            user.playSound(Sound.ENTITY_ENDERMAN_TELEPORT, 0.5f, 1f);

        }
    }
}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/gui/ArenaGUI.kt
package com.dumbdogdiner.warrior.gui

import com.dumbdogdiner.stickyapi.bukkit.gui.GUI
import com.dumbdogdiner.warrior.api.builders.ItemBuilder
import com.dumbdogdiner.stickyapi.bukkit.gui.ClickableSlot
import com.dumbdogdiner.warrior.api.arena.Arena
import java.util.Arrays
import com.dumbdogdiner.warrior.api.translation.Symbols
import com.dumbdogdiner.warrior.api.sessions.ArenaSession
import com.dumbdogdiner.warrior.Warrior
import com.dumbdogdiner.warrior.api.translation.Constants
import java.util.HashMap
import com.dumbdogdiner.warrior.api.util.TranslationUtil
import com.dumbdogdiner.warrior.user.User
import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryOpenEvent

class ArenaGUI : GUI(5, "Select Arena", Warrior.instance) {

    override fun onInventoryOpen(event: InventoryOpenEvent) {
        val placeholder = ItemBuilder(Material.BLACK_STAINED_GLASS_PANE).setName(" ").build()
        val blank = ItemBuilder(Material.GRAY_DYE).setName("§7Arena §3???").build()

        for (i in 0..8) {
            addSlot(i, 0, placeholder)
            addSlot(i, 4, placeholder)
        }

        val exit = ClickableSlot(Material.BARRIER, 1, "§c§lClose Menu", 4, 4)
        addSlot(exit)

        val arenaCount: Int = Warrior.arenaManager.arenas.size
        var i = 0
        for (y in 1..3) {
            for (x in 1..7) {
                if (i >= arenaCount) {
                    addSlot(x, y, blank)
                    continue
                }

                val a: Arena = Warrior.arenaManager.arenas[i]
                val avgRating = a.metadata.averageRating().toInt()

                val colorDesc = Arrays.stream(a.metadata.desc)
                    .map { str: String -> "§7§o$str" }
                    .toArray<String> { length -> arrayOfNulls(length) }

                val arenaItem = ItemBuilder(Material.FILLED_MAP)
                    .setName("§7Arena §3" + a.name)
                    .setLore(" ", "&7Rating &8» &b" + calcRating(avgRating), " ")
                    .appendLore(*colorDesc)
                    .build()

                val cs = ClickableSlot(arenaItem, x, y)
                addSlot(cs)

                i++
            }
        }
    }

    private fun calcRating(rating: Int): String {
        val stars = Symbols.BLACK_STAR * rating
        return stars + Symbols.WHITE_STAR * (5 - rating)
    }

    private operator fun Char.times(amount: Int): String = this.toString().repeat(amount)

    override fun onInventoryClick(event: InventoryClickEvent, tag: String?) {
        val item = event.currentItem ?: return

        val name = item.itemMeta.displayName

        if (name.contains("Close Menu")) {
            event.whoClicked.closeInventory()
        } else if (name.contains("Arena")) {
            val arenaName = name.replace("§7Arena §3", "")
            val user: User = Warrior.userCache[event.whoClicked.uniqueId]
            val a: Arena = Warrior.arenaManager[arenaName]

            user.session = ArenaSession(user.userId, a)

            val msg: String = Warrior.translator.translate(Constants.Lang.ARENA_TELEPORT, object : HashMap<String, String>() {
                init {
                    put("ARENA", a.name)
                }
            }, user)

            user.sendMessage(TranslationUtil.getPrefix() + msg)
            user.playSound(Sound.ENTITY_ENDERMAN_TELEPORT, 0.5f, 1f)
        }
    }
}
